src/sexpr.rs

```diff
@@ -213,22 +213,114 @@ impl Arena {
             unreachable!()
         }
     }
+
+    pub fn get_atom(&self, expr: SExpr) -> Option<&str> {
+        if !expr.is_atom() {
+            return None;
+        }
+
+        let inner = self.0.borrow();
+        // Safety argument: the data will live as long as the containing context, and is
+        // immutable once it's inserted, so using the lifteime of the Arena is acceptable.
+        let data = unsafe { std::mem::transmute(inner.strings[expr.index()].as_str()) };
+        Some(data)
+    }
+
+    pub fn get_str(&self, expr: SExpr) -> Option<&str> {
+        if !expr.is_string() {
+            return None;
+        }
+
+        let inner = self.0.borrow();
+        // Safety argument: the data will live as long as the containing context, and is
+        // immutable once it's inserted, so using the lifteime of the Arena is acceptable.
+        let data = unsafe { std::mem::transmute(inner.strings[expr.index()].as_str()) };
+        Some(data)
+    }
+
+    pub fn get_list(&self, expr: SExpr) -> Option<&[SExpr]> {
+        if !expr.is_list() {
+            return None;
+        }
+
+        let inner = self.0.borrow();
+        // Safety argument: the data will live as long as the containing context, and is
+        // immutable once it's inserted, so using the lifteime of the Arena is acceptable.
+        let data = unsafe { std::mem::transmute(inner.lists[expr.index()].as_slice()) };
+        return Some(data);
+    }
+
+    pub(crate) fn get_t<T: TryParseInt>(&self, expr: SExpr) -> Option<T> {
+        let inner = self.0.borrow();
+
+        if expr.is_atom() {
+            let data = inner.strings[expr.index()].as_str();
+            return T::try_parse_t(data, false);
+        }
+
+        if expr.is_list() {
+            let data = inner.lists[expr.index()].as_slice();
+
+            if data.len() != 2 || data.iter().any(|expr| !expr.is_atom()) {
+                return None;
+            }
+
+            let is_negated = inner.strings[data[0].index()].as_str() == "-";
```

Comment:
```suggestion
            let is_negated = match inner.strings[data[0].index()].as_str() {
                "-" => true,
                "+" => false,
                _ => return None,
            };
```

Otherwise we would incorrectly parse `(foo 123)` as `123`, right?